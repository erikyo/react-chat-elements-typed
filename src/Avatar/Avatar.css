.rce-avatar-container {
    position: relative;
    display: flex;
    justify-content: center;
    align-items: center;
}

.rce-avatar-container .rce-avatar {
    width: 100%;
    height: 100%;
    overflow: hidden;
}

.rce-avatar-container.rounded .rce-avatar {
    border-radius: 100%;
}

.rce-avatar-container.flexible .rce-avatar {
    height: auto !important;
    width: 100% !important;
    border-radius: unset !important;
    overflow: unset !important;
}

.rce-avatar-container.default {
    width: 50px;
    height: 50px;
}

.rce-avatar-container.xsmall {
    width: 20px;
    height: 20px;
}

.rce-avatar-container.small {
    width: 40px;
    height: 40px;
}

.rce-avatar-container.medium {
    width: 50px;
    height: 50px;
}

.rce-avatar-container.large {
    width: 64px;
    height: 64px;
}

.rce-avatar-container.xlarge {
    width: 100px;
    height: 100px;
}

.rce-avatar-container.rce-citem-avatar-encircle-status {
    box-sizing: border-box;
    position: relative;
}

.rce-avatar-letter-background {
    height: 100%;
    width: 100%;
}

.rce-avatar-letter {
    margin: 0;
    width: 100%;
    padding: 0 0 5%;
    color: var(--rce-color-white);
    height: 100%;
}
