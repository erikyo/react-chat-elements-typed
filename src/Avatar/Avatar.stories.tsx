import type { Meta, StoryObj } from "@storybook/react";

import Avatar from "./Avatar";
import { generateAvatar } from "../stories/utils";

const meta = {
	component: Avatar,
	render: (args) => (
		<div style={{ width: 250, height: 250 }}>
			<Avatar {...args} />
		</div>
	),
};

export default meta;

type Story = StoryObj<typeof meta>;

export const Default = {
	args: {
		...generateAvatar(24),
	},
};

export const small = {
	args: {
		...generateAvatar(1, { size: "small" }),
	},
};

export const big = {
	args: {
		...generateAvatar(1, { size: "xlarge" }),
	},
};
