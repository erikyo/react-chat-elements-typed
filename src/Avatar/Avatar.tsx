import type { FC } from "react";
import { useEffect } from "react";
import "./Avatar.css";
import classNames from "classnames";
import type { IAvatarProps } from "../type.js";

const Avatar: FC<IAvatarProps> = ({
	type = "default",
	size = "default",
	lazyLoadingImage = undefined,
	...props
}) => {
	const loadedAvatars: string[] = [];
	let loading = false;
	let src = props.src;
	let isLazyImage = false;

	useEffect(() => {
		if (lazyLoadingImage) {
			isLazyImage = true;

			if (!isLoaded(src)) {
				src = lazyLoadingImage;

				if (!loading) {
					requestImage(props.src);
				}
			} else {
				isLazyImage = false;
			}
		}
	}, [lazyLoadingImage, src, isLazyImage, loading, props.src]);

	const isLoaded = (src: string) => {
		return loadedAvatars.indexOf(src) !== -1;
	};

	const requestImage = (src: string) => {
		loading = true;

		const loaded = () => {
			loadedAvatars.push(src);
			loading = false;
		};

		const img: HTMLImageElement = document.createElement("img");
		img.src = src;
		img.onload = loaded;
		img.onerror = loaded;
	};

	const stringToColour = (str: string) => {
		let hash = 0;
		for (let i = 0; i < str.length; i++) {
			hash = str.charCodeAt(i) + ((hash << 5) - hash);
		}
		let colour = "#";
		for (let i = 0; i < 3; i++) {
			let value: number = (hash >> (i * 8)) & 0xff;
			value = (value % 150) + 50;
			colour += `00${value.toString(16)}`.substr(-2);
		}
		return colour;
	};

	return (
		<div
			className={classNames(
				"rce-avatar-container",
				type,
				size,
				props.className,
			)}
		>
			{props.letterItem ? (
				<div
					className="rce-avatar-letter-background"
					style={{ backgroundColor: stringToColour(props.letterItem.id) }}
				>
					<span className="rce-avatar-letter">{props.letterItem.letter}</span>
				</div>
			) : (
				<img
					alt={props.alt}
					src={src}
					onError={props.onError}
					className={classNames("rce-avatar", {
						"rce-avatar-lazy": isLazyImage,
					})}
				/>
			)}
			{props.sideElement}
		</div>
	);
};
export default Avatar;
