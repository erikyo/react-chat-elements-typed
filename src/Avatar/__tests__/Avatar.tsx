import { describe, expect, it } from "vitest";
import React from "react";
import Avatar from "../Avatar";
import { render } from "@testing-library/react";

describe("Avatar component", () => {
	it("should render without issues", () => {
		const component = render(<Avatar />);
		expect(component.container).toBeTruthy();
		expect(component.container).toMatchSnapshot();
	});
});
