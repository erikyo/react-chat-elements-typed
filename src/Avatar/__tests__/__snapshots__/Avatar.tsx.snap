// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Avatar component > should render without issues 1`] = `
<div>
  <div
    class="rce-avatar-container default rce-avatar rounded"
  >
    <img
      alt="Avatar"
      class="rce-avatar"
      src="data:image/svg+xml;utf8,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20viewBox%3D%220%200%2064%2064%22%20fill%3D%22none%22%20shape-rendering%3D%22auto%22%3E%3Cmetadata%20xmlns%3Ardf%3D%22http%3A%2F%2Fwww.w3.org%2F1999%2F02%2F22-rdf-syntax-ns%23%22%20xmlns%3Axsi%3D%22http%3A%2F%2Fwww.w3.org%2F2001%2FXMLSchema-instance%22%20xmlns%3Adc%3D%22http%3A%2F%2Fpurl.org%2Fdc%2Felements%2F1.1%2F%22%20xmlns%3Adcterms%3D%22http%3A%2F%2Fpurl.org%2Fdc%2Fterms%2F%22%3E%3Crdf%3ARDF%3E%3Crdf%3ADescription%3E%3Cdc%3Atitle%3EMiniavs%20-%20Free%20Avatar%20Creator%3C%2Fdc%3Atitle%3E%3Cdc%3Acreator%3EWebpixels%3C%2Fdc%3Acreator%3E%3Cdc%3Asource%20xsi%3Atype%3D%22dcterms%3AURI%22%3Ehttps%3A%2F%2Fwww.figma.com%2Fcommunity%2Ffile%2F923211396597067458%3C%2Fdc%3Asource%3E%3Cdcterms%3Alicense%20xsi%3Atype%3D%22dcterms%3AURI%22%3Ehttps%3A%2F%2Fcreativecommons.org%2Flicenses%2Fby%2F4.0%2F%3C%2Fdcterms%3Alicense%3E%3Cdc%3Arights%3ERemix%20of%20%E2%80%9EMiniavs%20-%20Free%20Avatar%20Creator%E2%80%9D%20(https%3A%2F%2Fwww.figma.com%2Fcommunity%2Ffile%2F923211396597067458)%20by%20%E2%80%9EWebpixels%E2%80%9D%2C%20licensed%20under%20%E2%80%9ECC%20BY%204.0%E2%80%9D%20(https%3A%2F%2Fcreativecommons.org%2Flicenses%2Fby%2F4.0%2F)%3C%2Fdc%3Arights%3E%3C%2Frdf%3ADescription%3E%3C%2Frdf%3ARDF%3E%3C%2Fmetadata%3E%3Cmask%20id%3D%22viewboxMask%22%3E%3Crect%20width%3D%2264%22%20height%3D%2264%22%20rx%3D%220%22%20ry%3D%220%22%20x%3D%220%22%20y%3D%220%22%20fill%3D%22%23fff%22%20%2F%3E%3C%2Fmask%3E%3Cg%20mask%3D%22url(%23viewboxMask)%22%3E%3Cpath%20d%3D%22M45.89%2036.1c0%208.5-1.26%2018.86-10.89%2019.82v9.95S31.36%2068%2026.5%2068c-4.86%200-8.5-3.48-8.5-3.48V42a5%205%200%200%201-1.3-9.83C15.36%2022.64%2017.5%2013%2032%2013c14.59%200%2014.24%2011.08%2013.96%2019.81-.04%201.15-.07%202.25-.07%203.29Z%22%20fill%3D%22%23ffcb7e%22%2F%3E%3Cpath%20d%3D%22M35%2055.92c-.48.05-.98.07-1.5.07-8.88%200-13.9-7.15-15.5-14.6v23.13S21.64%2068%2026.5%2068c4.86%200%208.5-2.13%208.5-2.13v-9.95Z%22%20fill%3D%22%23000%22%20fill-opacity%3D%22.07%22%2F%3E%3Cpath%20d%3D%22M34.63%2055.95c-.37.03-.74.04-1.13.04-6.53%200-10.97-3.86-13.5-8.87V48.24c0%205.38%202.61%209.75%208.28%209.75h1.35c3.34.03%204.59.04%205-2.04ZM16.7%2032.17A5%205%200%200%200%2018.14%2042c-.48-1.98-.71-3.99-.71-5.9a46.7%2046.7%200%200%201-.73-3.93Z%22%20fill%3D%22%23000%22%20fill-opacity%3D%22.07%22%2F%3E%3Crect%20x%3D%2236%22%20y%3D%2241%22%20width%3D%223%22%20height%3D%222%22%20rx%3D%221%22%20fill%3D%22%23000%22%20fill-opacity%3D%22.07%22%2F%3E%3Crect%20x%3D%227%22%20y%3D%2260%22%20width%3D%2240%22%20height%3D%2223%22%20rx%3D%229%22%20fill%3D%22%23ff4dd8%22%2F%3E%3Cpath%20d%3D%22M22%2028c-.63%203%201%206.98%201%207.74%200%20.77-3.93%203.03-5%203.76-1.07.73-1.5-7-1.5-7-3%200-3.5%205.5-3.5%205.5s-2.25-.74-3-4.5c-.51-2.54.3-8.09.5-9.5.5-3.5%201-11.5%207.5-15.5s23-4%2027-3C54.9%207.97%2056.22%2021.5%2053%2026c-5%205.5-19-1-23.5-1s-6.87%200-7.5%203Z%22%20fill%3D%22%2347280b%22%2F%3E%3Cg%20transform%3D%22translate(1)%22%3E%3Cpath%20d%3D%22M27.93%2046a1%201%200%200%201%201-1h9.14a1%201%200%200%201%201%201%205%205%200%200%201-5%205h-1.14a5%205%200%200%201-5-5Z%22%20fill%3D%22%2366253C%22%2F%3E%3Cpath%20d%3D%22M35.76%2050.7a5%205%200%200%201-1.69.3h-1.14a5%205%200%200%201-5-4.8c.77-.29%201.9-.25%203.02-.22L32%2046c2.21%200%204%201.57%204%203.5%200%20.42-.09.83-.24%201.2Z%22%20fill%3D%22%23B03E67%22%2F%3E%3Cpath%20d%3D%22M29%2045h10v1a1%201%200%200%201-1%201h-8a1%201%200%200%201-1-1v-1Z%22%20fill%3D%22%23fff%22%2F%3E%3Cpath%20d%3D%22M31%2045.3c0-.17.13-.3.3-.3h1.4c.17%200%20.3.13.3.3v2.4a.3.3%200%200%201-.3.3h-1.4a.3.3%200%200%201-.3-.3v-2.4Z%22%20fill%3D%22%23B03E67%22%2F%3E%3C%2Fg%3E%3Cg%20transform%3D%22translate(0%20-1)%22%3E%3Cpath%20d%3D%22M30%2037.5a1.5%201.5%200%200%201%203%200v1.23c0%20.15-.12.27-.27.27h-2.46a.27.27%200%200%201-.27-.27V37.5ZM40%2037.5a1.5%201.5%200%200%201%203%200v1.23c0%20.15-.12.27-.27.27h-2.46a.27.27%200%200%201-.27-.27V37.5Z%22%20fill%3D%22%231B0B47%22%2F%3E%3C%2Fg%3E%3C%2Fg%3E%3C%2Fsvg%3E"
    />
  </div>
</div>
`;
